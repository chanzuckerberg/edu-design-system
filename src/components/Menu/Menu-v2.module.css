@import '../../design-tokens/mixins.css';

/*------------------------------------*\
    # MENU
\*------------------------------------*/

/**
 * Menu
 */
.menu {
  position: relative;
}

.menu__button {
  font: var(--eds-theme-typography-body-md);

  color: var(--eds-theme-color-text-neutral-subtle);
  background-color: var(--eds-theme-color-form-background);
  border-color: var(--eds-theme-color-form-border);
  font-weight: var(--eds-font-weight-light);
}

.menu__button--with-chevron {
  color: var(--eds-theme-color-icon-neutral-default);
}

.menu__item {
  text-decoration: none;
  color: inherit;
}

/* Unset the hover on the menu item, as this is handled by the PopoverListItem */
.menu__item:hover {
  color: unset;
}