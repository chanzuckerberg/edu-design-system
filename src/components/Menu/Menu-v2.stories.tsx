import type { StoryObj, Meta } from '@storybook/react';
import { userEvent } from '@storybook/testing-library';
import React from 'react';

import { Menu } from './Menu-v2';
import type { MenuProps } from './Menu-v2';
import icons from '../../icons/spritemap';
import type { IconName } from '../../icons/spritemap';
import { Avatar } from '../Avatar/Avatar';

import { Icon } from '../Icon/Icon';
export default {
  title: 'Components/V2/Menu',
  component: Menu,
  parameters: {
    badges: ['intro-1.2', 'current-2.0'],
    layout: 'centered',
  },
  argTypes: {
    children: {
      control: {
        type: null,
      },
    },
  },
  decorators: [
    (Story) => (
      <div className="p-8">
        <Story />
      </div>
    ),
  ],
} as Meta<MenuProps>;

export const Default: StoryObj<MenuProps> = {
  args: {
    children: (
      <>
        <Menu.Button>Documentation Links</Menu.Button>
        <Menu.Items data-testid="menu-content">
          <Menu.Item
            href="https://headlessui.com/react/menu#menu-button"
            icon="link"
          >
            Headless UI Docs
          </Menu.Item>
          <Menu.Item
            href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/menu"
            icon="link"
          >
            MDN: Menu
          </Menu.Item>
          <Menu.Item href="#index" onClick={() => console.log('Item clicked')}>
            Trigger Action
          </Menu.Item>
          <Menu.Item disabled href="https://example.org/" icon="warning-filled">
            Not Possible (disabled)
          </Menu.Item>
        </Menu.Items>
      </>
    ),
  },
};

export const WithLongButtonText: StoryObj<MenuProps> = {
  args: {
    children: (
      <>
        <Menu.Button>
          Long Trigger Button Text to Demonstrate Popover Matching
        </Menu.Button>
        <Menu.Items data-testid="menu-content">
          <Menu.Item
            href="https://headlessui.com/react/menu#menu-button"
            icon="link"
          >
            Headless UI Docs
          </Menu.Item>
          <Menu.Item
            href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/menu"
            icon="link"
          >
            MDN: Menu
          </Menu.Item>
          {/* eslint-disable-next-line no-alert */}
          <Menu.Item onClick={() => alert('Item clicked')}>
            Trigger Action
          </Menu.Item>
          <Menu.Item disabled href="https://example.org/" icon="warning-filled">
            Not Possible (disabled)
          </Menu.Item>
        </Menu.Items>
      </>
    ),
  },
};

export const WithShortButtonText: StoryObj<MenuProps> = {
  args: {
    children: (
      <>
        <Menu.Button>Menu</Menu.Button>
        <Menu.Items data-testid="menu-content">
          <Menu.Item
            href="https://headlessui.com/react/menu#menu-button"
            icon="link"
          >
            Headless UI Docs
          </Menu.Item>
          <Menu.Item
            href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/menu"
            icon="link"
          >
            MDN: Menu
          </Menu.Item>
          {/* eslint-disable-next-line no-alert */}
          <Menu.Item onClick={() => alert('Item clicked')}>
            Trigger Action
          </Menu.Item>
          <Menu.Item disabled href="https://example.org/" icon="warning-filled">
            Not Possible (disabled)
          </Menu.Item>
        </Menu.Items>
      </>
    ),
  },
};

export const WithCustomButton: StoryObj<MenuProps> = {
  args: {
    children: (
      <>
        <Menu.PlainButton>
          <div className="fpo">Menu Button</div>
        </Menu.PlainButton>
        <Menu.Items data-testid="menu-content">
          <Menu.Item
            href="https://headlessui.com/react/menu#menu-button"
            icon="link"
          >
            Headless UI Docs
          </Menu.Item>
          <Menu.Item
            href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/menu"
            icon="link"
          >
            MDN: Menu
          </Menu.Item>
          {/* eslint-disable-next-line no-alert */}
          <Menu.Item onClick={() => alert('Item clicked')}>
            Trigger Action
          </Menu.Item>
          <Menu.Item disabled href="https://example.org/" icon="warning-filled">
            Not Possible (disabled)
          </Menu.Item>
        </Menu.Items>
      </>
    ),
  },
};

export const MenuWithAvatarButton: StoryObj<MenuProps> = {
  parameters: {
    badges: ['intro-1.3', 'implementationExample'],
  },
  args: {
    children: (
      <>
        <Menu.PlainButton>
          <Avatar user={{ fullName: 'Josie Sandberg' }} />
        </Menu.PlainButton>
        <Menu.Items data-testid="menu-content">
          <Menu.Item
            href="https://headlessui.com/react/menu#menu-button"
            icon="link"
          >
            Headless UI Docs
          </Menu.Item>
          <Menu.Item
            href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/menu"
            icon="link"
          >
            MDN: Menu
          </Menu.Item>
          {/* eslint-disable-next-line no-alert */}
          <Menu.Item onClick={() => alert('Item clicked')}>
            Trigger Action
          </Menu.Item>
          <Menu.Item disabled href="https://example.org/" icon="warning-filled">
            Not Possible (disabled)
          </Menu.Item>
        </Menu.Items>
      </>
    ),
  },
};

export const Opened: StoryObj<MenuProps> = {
  ...Default,
  parameters: {
    ...Default.parameters,
    // Sets the delay (in milliseconds) for a specific story.
    chromatic: { delay: 300 },
  },
  play: async () => {
    await userEvent.tab();
    await userEvent.keyboard(' ', { delay: 300 });
  },
};

export const MenuWithIconButton: StoryObj<MenuProps & { iconName: IconName }> =
  {
    argTypes: {
      iconName: {
        control: 'radio',
        options: Object.keys(icons),
      },
    },
    args: {
      iconName: 'dots-vertical',
    },
    parameters: {
      badges: ['intro-1.2', 'implementationExample'],
    },
    render: ({ iconName }) => (
      <Menu>
        <Menu.PlainButton>
          <Icon
            name={iconName}
            purpose="informative"
            size="2rem"
            title="show more"
          />
        </Menu.PlainButton>
        <Menu.Items data-testid="menu-content">
          <Menu.Item
            href="https://headlessui.com/react/menu#menu-button"
            icon="link"
          >
            Headless UI Docs
          </Menu.Item>
          <Menu.Item
            href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/menu"
            icon="link"
          >
            MDN: Menu
          </Menu.Item>
          {/* eslint-disable-next-line no-alert */}
          <Menu.Item onClick={() => alert('Item clicked')}>
            Trigger Action
          </Menu.Item>
          <Menu.Item disabled href="https://example.org/" icon="warning-filled">
            Not Possible (disabled)
          </Menu.Item>
        </Menu.Items>
      </Menu>
    ),
  };
