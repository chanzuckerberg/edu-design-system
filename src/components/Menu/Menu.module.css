@import '../../design-tokens/mixins.css';

/*------------------------------------*\
    # MENU
\*------------------------------------*/

/**
 * Menu
 */
.menu {
  position: relative;
}

.menu__button {
  font: var(--eds-theme-typography-body-md);

  color: var(--eds-theme-color-text-neutral-subtle);
  background-color: var(--eds-theme-color-form-background);
  border-color: var(--eds-theme-color-form-border);
  font-weight: var(--eds-font-weight-light);
}

.menu__plain-button {
  /* These min sizes are for target size accessibility */
  min-height: 44px;
  min-width: 44px;

  display: flex;
  justify-content: space-evenly;
  align-items: center;

  border-radius: var(--eds-border-radius-md);
}
.menu__plain-button:hover {
  background-color: var(
    --eds-theme-color-button-secondary-neutral-background-hover
  );
}
.menu__plain-button:focus-visible {
  @mixin focus;
}
@supports not selector(:focus-visible) {
  .menu__plain-button:focus {
    @mixin focus;
  }
}

.menu__button--with-chevron {
  color: var(--eds-theme-color-icon-neutral-default);
}

.menu__item {
  text-decoration: none;
  color: inherit;
}

/* Unset the hover on the menu item, as this is handled by the PopoverListItem */
.menu__item:hover {
  color: unset;
}