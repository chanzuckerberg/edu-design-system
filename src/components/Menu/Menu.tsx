import { Menu as HeadlessMenu } from '@headlessui/react';
import clsx from 'clsx';
import React from 'react';

import styles from './Menu.module.css';
import type { ExtractProps } from '../../util/utility-types';

import ClickableStyle from '../ClickableStyle';
import Icon from '../Icon';
import type { IconName } from '../Icon';
import PopoverContainer from '../PopoverContainer';
import PopoverListItem from '../PopoverListItem';

// Note: added className here to prevent private interface collision within HeadlessUI
export type MenuProps = ExtractProps<typeof HeadlessMenu> & {
  /**
   * Allow custom classes to be applied to the menu container.
   */
  className?: string;
};

export type MenuItemProps = ExtractProps<typeof HeadlessMenu.Item> & {
  href?: string;
  icon?: IconName;
  onClick?: () => any;
};

export type MenuButtonProps = ExtractProps<typeof HeadlessMenu.Button>;
export type MenuItemsProps = ExtractProps<typeof HeadlessMenu.Items>;

/**
 * BETA: This component is still a work in progress and is subject to change.
 *
 * `import {Menu} from "@chanzuckerberg/eds";`
 *
 * Creates a list of actions which appear next to a trigger component. This should
 * be used when there is a discrete number of actions such as user navigations (e.g.,
 * a profile menu with links to settings) or a set of actions
 */
export const Menu = (props: MenuProps) => {
  return <HeadlessMenu {...props} />;
};

/**
 * Trigger for the popover menu (dropdown)
 */
const MenuButton = ({ children, className, ...other }: MenuButtonProps) => {
  const buttonClassNames = clsx(styles['menu__button'], className);
  return (
    <HeadlessMenu.Button
      as={ClickableStyle}
      className={buttonClassNames}
      status="neutral"
      {...other}
    >
      {children}
      <Icon
        className={styles['menu__button--icon']}
        name="expand-more"
        purpose="decorative"
        size="1.5rem"
      />
    </HeadlessMenu.Button>
  );
};

/**
 * A set of menu items in the Menu
 *
 * @param props Props used on the set of menu items
 * @see https://headlessui.com/react/menu#menu-items
 */
const MenuItems = (props: MenuItemsProps) => (
  <HeadlessMenu.Items
    as={PopoverContainer}
    className={clsx(styles['menu__popover'])}
    data-testid="menu-content"
    {...props}
  />
);

/**
 * Individual menu items, styled with the popover list item component
 */
const MenuItem = ({
  children,
  href,
  icon,
  onClick,
  ...other
}: MenuItemProps) => {
  const destinationUrl = !onClick ? href : undefined;
  return (
    <HeadlessMenu.Item {...other}>
      {({ active, disabled }) => {
        const listItemView = (
          <PopoverListItem active={active} disabled={disabled} icon={icon}>
            {children}
          </PopoverListItem>
        );
        return disabled ? (
          listItemView
        ) : (
          <a
            className={clsx(styles['menu__item'])}
            href={destinationUrl}
            onClick={onClick}
          >
            {listItemView}
          </a>
        );
      }}
    </HeadlessMenu.Item>
  );
};

Menu.Button = MenuButton;
Menu.Items = MenuItems;
Menu.Item = MenuItem;
