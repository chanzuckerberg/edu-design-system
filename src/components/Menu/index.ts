export { Menu as default } from './Menu';
export { Menu as MenuV2 } from './Menu-v2';
