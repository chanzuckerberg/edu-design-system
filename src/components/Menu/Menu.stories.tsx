import type { StoryObj, Meta } from '@storybook/react';
import React from 'react';

import { Menu } from './Menu';
import type { MenuProps } from './Menu';

export default {
  title: 'Components/Menu',
  component: Menu,
  subcomponents: {
    'Menu.Button': Menu.Button,
    'Menu.Items': Menu.Items,
    'Menu.Item': Menu.Item,
  },
  parameters: {
    badges: ['1.2'],
    layout: 'centered',
  },
} as Meta<MenuProps>;

export const Default: StoryObj<MenuProps> = {
  args: {
    children: (
      <>
        <Menu.Button>Documentation Links</Menu.Button>
        <Menu.Items data-testid="menu-content">
          <Menu.Item
            href="https://headlessui.com/react/menu#menu-button"
            icon="link"
          >
            Headless UI Docs
          </Menu.Item>
          <Menu.Item
            href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/menu"
            icon="link"
          >
            MDN: Menu
          </Menu.Item>
          {/* eslint-disable-next-line no-alert */}
          <Menu.Item onClick={() => alert('Item clicked')}>
            Trigger Action
          </Menu.Item>
          <Menu.Item disabled href="https://example.org/" icon="warning">
            Not Possible (disabled)
          </Menu.Item>
        </Menu.Items>
      </>
    ),
  },
};

export const WithLongButtonText: StoryObj<MenuProps> = {
  args: {
    children: (
      <>
        <Menu.Button>
          Long Trigger Button Text to Demonstrate Popover Matching
        </Menu.Button>
        <Menu.Items data-testid="menu-content">
          <Menu.Item
            href="https://headlessui.com/react/menu#menu-button"
            icon="link"
          >
            Headless UI Docs
          </Menu.Item>
          <Menu.Item
            href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/menu"
            icon="link"
          >
            MDN: Menu
          </Menu.Item>
          {/* eslint-disable-next-line no-alert */}
          <Menu.Item onClick={() => alert('Item clicked')}>
            Trigger Action
          </Menu.Item>
          <Menu.Item disabled href="https://example.org/" icon="warning">
            Not Possible (disabled)
          </Menu.Item>
        </Menu.Items>
      </>
    ),
  },
};

export const WithShortButtonText: StoryObj<MenuProps> = {
  args: {
    children: (
      <>
        <Menu.Button>Menu</Menu.Button>
        <Menu.Items data-testid="menu-content">
          <Menu.Item
            href="https://headlessui.com/react/menu#menu-button"
            icon="link"
          >
            Headless UI Docs
          </Menu.Item>
          <Menu.Item
            href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/menu"
            icon="link"
          >
            MDN: Menu
          </Menu.Item>
          {/* eslint-disable-next-line no-alert */}
          <Menu.Item onClick={() => alert('Item clicked')}>
            Trigger Action
          </Menu.Item>
          <Menu.Item disabled href="https://example.org/" icon="warning">
            Not Possible (disabled)
          </Menu.Item>
        </Menu.Items>
      </>
    ),
  },
};

export const WithCustomButton: StoryObj<MenuProps> = {
  args: {
    children: (
      <>
        <Menu.PlainButton>
          <div className="fpo">Menu Button</div>
        </Menu.PlainButton>
        <Menu.Items data-testid="menu-content">
          <Menu.Item
            href="https://headlessui.com/react/menu#menu-button"
            icon="link"
          >
            Headless UI Docs
          </Menu.Item>
          <Menu.Item
            href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/menu"
            icon="link"
          >
            MDN: Menu
          </Menu.Item>
          {/* eslint-disable-next-line no-alert */}
          <Menu.Item onClick={() => alert('Item clicked')}>
            Trigger Action
          </Menu.Item>
          <Menu.Item disabled href="https://example.org/" icon="warning">
            Not Possible (disabled)
          </Menu.Item>
        </Menu.Items>
      </>
    ),
  },
};
