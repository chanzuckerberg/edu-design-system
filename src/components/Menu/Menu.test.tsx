import { generateSnapshots } from '@chanzuckerberg/story-utils';
import { userEvent } from '@storybook/testing-library';
import { composeStories } from '@storybook/testing-react';
import { render, screen } from '@testing-library/react';

import React from 'react';
import * as stories from './Menu.stories';

const { Default } = composeStories(stories);

describe('<Menu />', () => {
  generateSnapshots(stories, {
    getElement: async () => {
      const triggerButton = await screen.findByRole('button');
      userEvent.click(triggerButton);
      return triggerButton.parentElement; // eslint-disable-line testing-library/no-node-access
    },
  });

  it('should allow for keyboard navigation to enabled menu items', async () => {
    render(<Default />);
    const triggerButton = await screen.findByRole('button');

    userEvent.click(triggerButton);
    const menuContainer = await screen.findByRole('menu');

    // a11y requires attaching the control dynamically, so confirm
    expect(triggerButton.getAttribute('aria-controls')).toEqual(
      menuContainer.getAttribute('id'),
    );

    expect(menuContainer.getAttribute('aria-activedescendant')).toBeNull();

    userEvent.keyboard('{arrowdown}');

    expect(menuContainer.getAttribute('aria-activedescendant')).not.toBeNull();
  });

  it('should close menu on keyboard escape key', async () => {
    render(<Default />);
    const triggerButton = await screen.findByRole('button');

    userEvent.click(triggerButton);
    const menuContainer = await screen.findByRole('menu');

    // a11y attaches the control dynamically, so confirm
    expect(triggerButton.getAttribute('aria-controls')).toEqual(
      menuContainer.getAttribute('id'),
    );

    userEvent.keyboard('{esc}');

    expect(screen.queryByTestId('menu-content')).not.toBeInTheDocument();
  });
});
