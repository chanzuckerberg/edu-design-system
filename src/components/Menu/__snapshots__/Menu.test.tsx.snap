// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> Default story renders snapshot 1`] = `
<div>
  <button
    aria-controls="headlessui-menu-items-2"
    aria-expanded="true"
    aria-haspopup="true"
    class="clickable-style clickable-style--lg clickable-style--secondary clickable-style--neutral menu__button"
    data-bootstrap-override="clickable-style-secondary"
    id="headlessui-menu-button-1"
    type="button"
  >
    Documentation Links
    <svg
      aria-hidden="true"
      class="icon menu__button--icon"
      fill="currentColor"
      height="1.5rem"
      style="--icon-size: 1.5rem;"
      width="1.5rem"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="test-file-stub#expand-more"
      />
    </svg>
  </button>
  <div
    aria-labelledby="headlessui-menu-button-1"
    class="popover-container menu__popover"
    data-testid="menu-content"
    id="headlessui-menu-items-2"
    role="menu"
    tabindex="0"
  >
    <a
      class="menu__item"
      href="https://headlessui.com/react/menu#menu-button"
      id="headlessui-menu-item-3"
      role="menuitem"
      tabindex="-1"
    >
      <div
        class="popover-list-item"
      >
        <div
          class="popover-list-item__icon"
        >
          <svg
            aria-hidden="true"
            class="icon"
            fill="currentColor"
            height="1.5rem"
            style="--icon-size: 1.5rem;"
            width="1.5rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="test-file-stub#link"
            />
          </svg>
        </div>
        Headless UI Docs
      </div>
    </a>
    <a
      class="menu__item"
      href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/menu"
      id="headlessui-menu-item-4"
      role="menuitem"
      tabindex="-1"
    >
      <div
        class="popover-list-item"
      >
        <div
          class="popover-list-item__icon"
        >
          <svg
            aria-hidden="true"
            class="icon"
            fill="currentColor"
            height="1.5rem"
            style="--icon-size: 1.5rem;"
            width="1.5rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="test-file-stub#link"
            />
          </svg>
        </div>
        MDN: Menu
      </div>
    </a>
    <a
      class="menu__item"
      id="headlessui-menu-item-5"
      role="menuitem"
      tabindex="-1"
    >
      <div
        class="popover-list-item"
      >
        <div
          class="popover-list-item__no-icon"
        />
        Trigger Action
      </div>
    </a>
    <div
      aria-disabled="true"
      class="popover-list-item popover-list-item--disabled"
      id="headlessui-menu-item-6"
      role="menuitem"
    >
      <div
        class="popover-list-item__icon"
      >
        <svg
          aria-hidden="true"
          class="icon"
          fill="currentColor"
          height="1.5rem"
          style="--icon-size: 1.5rem;"
          width="1.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="test-file-stub#warning"
          />
        </svg>
      </div>
      Not Possible (disabled)
    </div>
  </div>
</div>
`;
